src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -386,101 +430,125 @@ def zone_op(self):
+            secret_key = sys_user.keys[0].secret_key if sys_user and sys_user.keys else ''
+            sys_user.add_key(access_key, secret_key)
+            return sys_user
         except RGWAMException as e:
             raise RGWAMException('failed to create system user', e)
 
-        sys_user = RGWUser(sys_user_info)
-
-        logging.info('Created system user: %s' % sys_user.uid)
-
-        sys_access_key = ''
-        sys_secret = ''
-
-        if len(sys_user.keys) > 0:
-            sys_access_key = sys_user.keys[0].access_key
-            sys_secret = sys_user.keys[0].secret_key
-
-        try:
-            zone_info = self.zone_op().modify(zone, zg, endpoints, None, sys_access_key, sys_secret)
-        except RGWAMException as e:
-            raise RGWAMException('failed to modify zone info', e)
-
+    def create_normal_user(self, zg, zone, uid=None):
         try:
             user_info = self.user_op().create(zone, zg, uid=uid, is_system=False)
+            user = RGWUser(user_info)
+            logging.info('Created regular user: %s' % user.uid)
+            return user
         except RGWAMException as e:
             raise RGWAMException('failed to create user', e)
 
-        user = RGWUser(user_info)
-
-        logging.info('Created regular user: %s' % user.uid)
-
-        eps = endpoints.split(',')
-        ep = ''
-        if len(eps) > 0:
-            ep = eps[0]
-            if start_radosgw:
-                o = urlparse(ep)
-                svc_id = realm_name + '.' + zone_name
-                self.env.mgr.apply_rgw(svc_id, realm_name, zone_name, o.port)
-
-        realm_token = RealmToken(realm_id, ep, sys_user.uid, sys_access_key, sys_secret)
-
-        logging.info(realm_token.to_json())
+    def update_period(self, realm, zg, zone=None):
+        try:
+            period_info = self.period_op().update(realm, zg, zone, commit=True)
+            period = RGWPeriod(period_info)
+            logging.info('Period: ' + period.id)
+        except RGWAMCmdRunException as e:
+            raise RGWAMException('failed to update period', e)
 
-        realm_token_b = realm_token.to_json().encode('utf-8')
-        return (0, 'Realm Token: %s' % base64.b64encode(realm_token_b).decode('utf-8'), '')
+    def realm_bootstrap(self, rgw_spec, start_radosgw=True):
+
+        realm_name = rgw_spec.rgw_realm
+        zonegroup_name = rgw_spec.rgw_zonegroup
+        zone_name = rgw_spec.rgw_zone
+
+        # Some sanity checks
+        if realm_name in self.realm_op().list():
+            raise RGWAMException(f'Realm {realm_name} already exists')
+        if zonegroup_name in self.zonegroup_op().list():
+            raise RGWAMException(f'ZonegroupOp {zonegroup_name} already exists')
+        if zone_name in self.zone_op().list():
+            raise RGWAMException(f'Zone {zone_name} already exists')
+
+        # Create RGW multisite entities and update the period
```

Comment:
is there something being done here specific to multisite or can this bootstrapped realm be used for single site as well?

Reply:
Normally the realm/zonegroup/zone creation is the same for both single and multisite configuration. I mean, AFAIK multi-site in fact only makes sense when you create a secondary zone on another cluster and you configure it to sync with the primary zone.

Reply:
okay, was just asking as the comment specified "multisite entitities". Sounds like they are general use though which is nice.

Reply:
I removed the mention to multisite from the comment since this is not only limited to the multisite case.